### PR TITLE
pick non null out of new and old values

### PR DESCRIPTION
**Pull request by sravfeyn**
Fix for  https://sentry.io/organizations/dimagi/issues/1429050750/?project=136860&query=is%3Aunresolved and https://github.com/dimagi/commcare-hq/pull/26305#issuecomment-575905734

@dannyroberts 

From the data it looks like, sometimes in the heartbeat view a sync-date may not be passed on

```
In [7]: UserReportingMetadataStaging.objects.filter(last_heartbeat__isnull=False, sync_date__isnull=True, received_on
   ...: __isnull=True).count()
Out[7]: 137

In [8]: UserReportingMetadataStaging.objects.filter(last_heartbeat__isnull=False, sync_date__isnull=True, received_on
   ...: __isnull=False).count()
Out[8]: 208
```

### CODE REVIEW

**Review by snopoke — APPROVED** [review]

**Review by dannyroberts — APPROVED** [review]

File: corehq/apps/users/util.py [hunk 1]
```diff
@@ -293,6 +293,8 @@ def update_last_sync(user, app_id, sync_date, version):
 
 
 def _last_sync_needs_update(last_sync, sync_datetime):
+    if not sync_datetime:
+        return False
```
Comment (dannyroberts) [inline corehq/apps/users/util.py]: I'm a bit on the fence about this line. Do we ever expect `sync_datetime` to be falsy now that it's always being set to something? I'm worried this line will essentially silently swallow any future errors. For example, if this had already been here before today, we never would have gotten the errors that led you to find that heartbeat wasn't always adding the sync_date correctly, right?

If what I'm saying is on the right track, maybe it's worth reporting the error to sentry so it's still visible that something unexpected is going on.

Comment (sravfeyn) [reply corehq/apps/users/util.py]: It's a good point, made me think more and indeed I found the bug that I introduced in https://github.com/dimagi/commcare-hq/pull/26305/. I made fixes for that here now.

File: corehq/apps/users/tasks.py [hunk 2]
```diff
@@ -338,10 +338,17 @@ def process_reporting_metadata_staging():
                     num_unsent_forms=record.num_unsent_forms,
                     num_quarantined_forms=record.num_quarantined_forms
                 )
+                if not record.last_heartbeat:
+                    # coming from sync
+                    latest_build_date = record.sync_date
+                else:
+                    # coming from hearbeat
+                    latest_build_date = record.modified_on
```
Comment (sravfeyn) [inline corehq/apps/users/tasks.py]: This now [reflects the code](https://github.com/dimagi/commcare-hq/pull/26305/files#diff-291dfd3802bf0daa51e598788d5ca13bL343) as it was before I modified it in this PR

File: corehq/pillows/synclog.py [hunk 3]
```diff
@@ -104,18 +104,22 @@ def process_change(self, change):
...
+        build_update_date=None):
     version = None
     if build_id:
         version = get_version_from_build_id(domain, build_id)
 
-    local_save = update_last_sync(user, app_id, sync_date, version)
+    local_save = False
+    if sync_date:
+        # sync_date could be null if this is called from a heartbeat request
+        update_last_sync(user, app_id, sync_date, version)
```
Comment (sravfeyn) [inline corehq/pillows/synclog.py]: This used to get called only if a sync_date was available. See https://github.com/dimagi/commcare-hq/pull/26305/files#diff-291dfd3802bf0daa51e598788d5ca13bL352

File: corehq/apps/users/util.py (hunk 1 above)
Comment (sravfeyn) [reply corehq/apps/users/util.py]: Also reverted this change, which is good to have for the reasons you have stated. Thanks!

**Review by snopoke — APPROVED** [review]

**Comment by dannyroberts** [comment]
Awesome. Looks like there are some related test failures, but hopefully nothing fundamental!

**Comment by dannyroberts** [comment]
Looks like it still has issues
```
======================================================================
ERROR: testapps.test_pillowtop.tests.test_xform_pillow:XFormPillowTest.test_app_metadata_tracker_heartbeat_processed
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python3.6/site-packages/django/test/utils.py", line 384, in inner
    return func(*args, **kwargs)
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/tests/test_xform_pillow.py", line 191, in test_app_metadata_tracker_heartbeat_processed
    self._test_heartbeat(0)
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/tests/test_xform_pillow.py", line 170, in _test_heartbeat
    230, 2, 10, 'CommCare 2.28', 'en'
  File "/mnt/commcare-hq-ro/corehq/apps/users/models.py", line 2912, in add_heartbeat
    """, params)
  File "/vendor/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/vendor/lib/python3.6/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/vendor/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/vendor/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: syntax error at or near "build_id"
LINE 4:                     build_id,
                            ^
```
How can I be playing a better role in this, if at all? Would it be less pressure if I revert something to stop the errors on production, and if so which PR should I be reverting at this point? I'm unfortunately not super familiar with the code and don't have a good sense of what the impact has been of having the issue be live for so long in production—do you think (if you were me) that I should be more concerned or less?

Thanks for working on this

**Comment by sravfeyn** [comment]
Hey @dannyroberts the build is passing now, I also made one small change as well. This should be good to be merged now, so may be a revert is not necessary.

**Comment by dannyroberts** [comment]
Thanks again @sravfeyn. I'm deploying this today, hopefully it fixes the issue

**Comment by dannyroberts** [comment]
Looks like this maybe caused these `ResourceConflicts`? https://sentry.io/organizations/dimagi/issues/1471476635/?environment=production&project=136860&query=is%3Aunresolved

**Comment by sravfeyn** [comment]
Fix here https://github.com/dimagi/commcare-hq/pull/26499

**Pull request merged**